### ISSUE TITLE
[Feature request] allows translations

### ISSUE THREAD

**Issue by trebmuh** [Manuel-K/shruthi-editor#11]
Hi, I'd very like to propose a translation in French for shruthi-editor.
It looks that the code doesn't allow to do so, so far.
I don't know how to do it, but I know it should be pretty straight forward thanks to using the Qt toolkit.

Could you please make it translatable then I could happily provide a French translation file?

Thank you.

_pinging @jpcima in case of... :wink:_

**Comment by Manuel-K** [comment]
I like the idea of translations, but there is some additional work required.

I've set the language of the `*.ui` files and marked texts that are automatically set with `notr`. That part has been pushed to master. 

I've modified the rest of the code base to allow translation of the names of the `patch` parameters and their values that are defined in `labels`. There are 360 unique translatable strings.

I'm not happy with @jpcima's `QTranslation` loading code or my quick and dirty implementation. I think the translations files should be searched in the application's directory and in `QStandardPaths::AppDataLocation` (maybe even in `QStandardPaths::DataLocation` as a fallback). So far the editor is backwards compatible with Qt4 and that does not have the `QStandardPaths` class.

**Comment by jpcima** [comment]
I have fixed the build of Qt4, and checked that it worked.
I have not used the class `QStandardPaths`; I have used `QLibraryInfo` which is available.

About the application search path; I expect the software to be able to find `qm` files regardless of platform, as long as the installation is laid out in FHS-style paths.
It's a possibility to organize binary bundles in this manner.

If it's preferred to have a specific way according to platforms rather than this, it can have special implementation.

**Comment by Manuel-K** [comment]
Yes, `QLibraryInfo` can be used to find the internal translation files for Qt. But the editors translation files should be searched in the paths specified by `QStandardPaths::AppDataLocation`, i.e. `.local/share/...`, `/usr/local/share/...` and `/usr/share/` on Linux and the corresponding paths on Windows/Mac. I don't want to add any hardcoded paths.

**Comment by jpcima** [comment]
Sorry to ask about this, I don't understand the request at all.
If I introduce `QStandardPaths`, the software loses the compatibility with Qt 4.

Secondly, the code does not hardcode a full path for translations. It is a relative path taken according of the location of the executable.
Various software have produced portable software directories by the same technique.

**Comment by Manuel-K** [comment]
Translations are implemented. I may have missed a few spots.

Translation files are embedded into the exe, but external files in the the exe's directory are given a higher priority to enable testing of translations without recompilation. As there does not seem to be an easy way (like `LC_ALL=C` on Linux) to force programs to start with a specific locale on Windows, I've added an option to disable translations to the settings dialog.

**Comment by jpcima** [comment]
Thanks for the work. I'll be able to complete the rest of the translation.
There will be a problem: configurations have varying success with this build, depending on Qt version.

The configs `lrelease` and `embed_translations` have no effect prior to [Qt 5.12](https://github.com/qt/qtbase/commit/01d2f35b6988edacc741b358703d16d4cd8c199a#diff-0e5a174e7dafaad904b9cb7af5a4c89c).
Debian stretch has Qt 5.7, and @trebmuh will be not be able to package this.

**Comment by Manuel-K** [comment]
I'm aware of that. I was hoping you would rebase your PR on master and adapt the qmake changes without including the partial translation.

I've got a partial German translation and I think we should create a separate branch for the unfinished translations.
